### DISCUSSION TITLE
OpenGL version is too old error

### DISCUSSION THREAD

**Discussion by LoopControl** [ghostty-org/ghostty#3159] [Q&A] [answered]
On Ubuntu 22, after building with the recommended Zig version, I'm getting the following message on startup (and it fails to start):

```
info: ghostty version=1.0.0-dev+0000000
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.6.9 runtime=4.6.9
info: optional config file not found, not loading path=/home/deity/.config/ghostty/config
info(config): default shell source=env value=/usr/bin/zsh
"vulkan-disable" is only available when building GTK with G_ENABLE_DEBUG. See GDK_DEBUG=help
info(gtk): libadwaita version build=1.1.7 runtime=1.1.7
error(gtk): unable to get current color scheme: GDBus.Error:org.freedesktop.DBus.Error.UnknownMethod: No such method “ReadOne”
info(grid): loaded OpenGL 3.2
warning(grid): OpenGL version is too old. Ghostty requires OpenGL 3.3
error(gtk_surface): surface failed to realize: error.OpenGLOutdated
[1]    2094758 segmentation fault (core dumped)  ghostty
```

Glxinfo shows my Nvidia card running recent drivers with openGL 4.6:
```
> glxinfo | grep "OpenGL version" 
OpenGL version string: 4.6.0 NVIDIA 550.127.05
```

Any suggestions?

Edit: The full GLXinfo does show an OpenGL ES version of 3.2:
```
OpenGL ES profile version string: OpenGL ES 3.2 NVIDIA 550.127.05
```

I'm not sure that this can be upgraded though as I'm already running the latest Nvidia drivers?

**Comment by kc8** [comment]
I ran into the same issue. I re-compiled with a glfw back end `zig build -Doptimize=ReleaseFast -Dapp-runtime=glfw` and ghostty compiled and seems to be working.

**Reply by mitchellh** [reply]
Note the GLFW build has bugs, memory corruption, crashes, and is missing dozens of features. It’s only used for Dev and isn’t suitable for daily use.

**Reply by LoopControl** [reply]
@mitchellh Since OpenGL ES 3.3 isn’t available for nvidia users on Linux I think it’ll have to do unless ES 3.2 is supported. 

@kc8 thanks I’ll test it out. Edit: No luck. It did compile and got a little further than the other build since I see the app in my taskbar with the ghost icon but no actual terminal (same complaint about opengl 3.3).

**Reply by nodonoghue** [reply]
Adding `-Dapp-runtime=glfw` fixed this issue for me.  This may be related to the system I had this issue with is running in hybrid graphics mode, doesn't seem that either the nvidia driver was being properly detected or it was using the intel driver instead of the discrete GPU.  This system is running Pop_OS.

I did not have this issue on my main desktop which has only a discrete graphics card, running EndeavorOS.

**Reply by rhodes-b** [reply]
Yeah this does seem related to detection since opengl 3.3 is like 15 years old so I would expect that bascially everything supports it at this point, OP also seems to have old GTK and adwaita versions not sure how much if any problem that would cause

**Reply by LoopControl** [reply]
> opengl 3.3 is like 15 years old

It seems ghostty is requiring OpenGL **ES** 3.3 which is actually not even released/standardized. 3.2 is latest ES version ( https://en.m.wikipedia.org/wiki/OpenGL_ES )

**Reply by mitchellh** [reply]
https://github.com/ghostty-org/ghostty/blob/bee21880149129f9373a35e9df0d39c223481709/src/apprt/gtk/Surface.zig#L438-L441

No, we explicit tell GTK we don't support ES. This is a GTK bug.

**Reply by zvrkSam** [reply]
Yeah adding `-Dapp-runtime=glfw` does the trick for now. I am also on Pop_OS with a nvidia card

**Reply by mitchellh** [reply]
Please heed my warning here:

> Note the GLFW build has bugs, memory corruption, crashes, and is missing dozens of features. It’s only used for Dev and isn’t suitable for daily use.

GLFW _will_ cause you problems. Its not a supported build.

**Reply by LoganGray** [reply]
> Yeah adding `-Dapp-runtime=glfw` does the trick for now. I am also on Pop_OS with a nvidia card

Exactly same situation here.   Please ping me and let me know if you find another way to do this since michellh is definitely saying GLFW is buggy.

**Reply by jbromley** [reply]
> This may be related to the system I had this issue with is running in hybrid graphics mode, doesn't seem that either the nvidia driver was being properly detected or it was using the intel driver instead of the discrete GPU. This system is running Pop_OS.
> 
> I did not have this issue on my main desktop which has only a discrete graphics card, running EndeavorOS.

That's interesting. I have a System 76 laptop with hybrid graphics and PopOS 22.04 and on that machine Ghostty works whether I'm in Intel or Nvidia graphics mode. It says it's loading OpenGL 4.6 and works fine. I've also got a desktop with an Nvidia card, and there Ghostty doesn't run -- it gives me the "OpenGL version is too old. Ghostty requires OpenGL 3.3" message. As far as I can tell, both machines have the same software since they were both provisioned from the same script. I've specifically checked gtk-4 packages, and all packages have the same version -- 4.6.9. From what I understand from this thread, it shouldn't work on either machine.

**Reply by nodonoghue** [reply]
I noticed this as well, the status, at least on Pop, has changed.  Either an update to Pop, or Ghostty resolved the OpenGL issue for me.

**Reply by marco-zan** [reply]
> I noticed this as well, the status, at least on Pop, has changed. Either an update to Pop, or Ghostty resolved the OpenGL issue for me.

Did you use nvidia or hybrid graphics mode? I found out that just with nvidia mode, ghostty doesn´t work. Maybe that's the reason it started working, not an update - mine seems to still be broken

**Reply by nodonoghue** [reply]
I'm running in hybrid mode, but I was when I was getting errors as well, so something definitely updated and fixed the issue for me.  I've been happily using Ghostty for a couple weeks now.

**Reply by alimoezzi** [reply]
FYI tabbing doesn't work with -Dapp-runtime=glfw
```
warning(glfw): tabbing is not supported on this platform
```

**Reply by rhodes-b** [reply]
FYI it has been stated many times GLFW is a debug only runtime (for developers of ghostty) and is missing a ton of features and has known bugs you should not use it for daily use

**Reply by 00-kat** [reply]
Note for future readers: the GLFW backend has since been removed because you were not meant to be using it; this purported solution no longer works.

**Comment by Calder-Ty** [comment]
For what it's worth, Updating to 1libgtk-4-dev version 4.14.2+ds-1ubuntu1 worked for me. This required updating my Ubuntu to 24.04, and rebuilding ghostty.

**Comment by Bananas-Are-Yellow** [comment, marked as answer]
The [Build from Source](https://ghostty.org/docs/install/build) instructions contain this note:

> A recent GTK is required for Ghostty to work with Nvidia (GL) drivers under x11. Ubuntu 22.04 LTS has GTK 4.6 which is not new enough. Ubuntu 23.10 has GTK 4.12 and works.

This is the reason for the problem with Ubuntu 22.04.

**Reply by nianfouyi** [reply]
Hi, sorry to bother you. I saw the information you mentioned, but I didn't quite understand. It means that Ubuntu 22.04 is not supported, right?

**Reply by marco-zan** [reply]
(I have no power or rights to say this whatsoever I am not a maintainer or anything, I just want to help purely looking at that comment and my experience)

In Ubuntu 22.04 with xorg and Nvidia it is not supported. 

It was working with pop os 22.04 in Wayland using the integrated Intel graphics but I never really used it there because I had other unrelated problems with the GPU itself

**Reply by jWeck-Software** [reply]
The bad thing is, anyone like me who still uses an old GTX and sees no reason to switch and doesn't want to spend the money is now f***ed.

Same with Warp!
> Your graphics card "GTX 650" and therefore driver is too old! You need to buy a new card that is supported by an up-to-date driver, otherwise you won't be able to use Warp.

Working with 4 monitors and all works just fine, I do not know why I should spend hundreds of € only to be able to use a **Terminal**!!

@LoopControl @ghostty-org @mitchellh

**Reply by 00-kat** [reply]
Because apparently terminals are complicated things. Try a terminal emulator that isn't GPU-accelerated.

Please note that bumping old threads (especially resolved ones) is strongly discouraged, as per the discussion rules → #6937.

Edit: and pinging everyone is *especially* crass. Did you even think that through for a *second* before you hit “Update comment”?

**Comment by Zorgatone** [comment]
I am on Ubuntu 26.04 (arm64 in a Parallels VM on macOS), and I also get this OpenGL version is too old error.

What version of Ubuntu should I be using?

**Reply by 00-kat** [reply]
It's probably not your Ubuntu version, but rather the OpenGL version your VM's emulated GPU claims to support. Consider opening a new discussion: I feel like the issue is somewhere in your Parallels configuration.

**Reply by Zorgatone** [reply]
Building from source produced a binary without errors, but running it produces the same OpenGL error.

VirtIO is enabled for hardware acceleration

**Reply by 00-kat** [reply]
As I said, open a new discussion.

**Edit**: moved to #12611.

**Comment by 00-kat** [comment]
This discussion is very old, so I'm going to lock it to prevent more comments here. If you still have issues, and it's not because your *graphics card* truly is too old, and the suggested answer here doesn't work, please [open a new discussion].

[open a new discussion]: https://github.com/ghostty-org/ghostty/discussions/new?category=issue-triage
